File: python/automation_agent/obs/log.py
"""Log <-> trace correlation for the obs package.

A logging filter that stamps every record emitted while a span is active with that span's
``trace_id`` and ``span_id``. This lets a backend pivot from a log line to the trace it
belongs to (and, on the GCP path, Cloud Logging auto-links the two). It reads the active span
from the current context, so it is zero-cost when no span is active or tracing is off — the
record's ids are then the empty string, and a formatter that references them never fails.
"""

from __future__ import annotations

import logging

from opentelemetry import trace


class TraceCorrelationFilter(logging.Filter):
    """A logging filter that adds ``trace_id`` / ``span_id`` to each record from the active
    span. When no span is active — or tracing is disabled, in which case the active span is
    the framework's no-op span with an invalid context — both are set to the empty string, so
    a format string that includes them is always safe. It never drops a record (``filter``
    always returns True)."""

    def filter(self, record: logging.LogRecord) -> bool:
        sc = trace.get_current_span().get_span_context()
        if sc.is_valid:
            record.trace_id = trace.format_trace_id(sc.trace_id)
            record.span_id = trace.format_span_id(sc.span_id)
        else:
            record.trace_id = ""
            record.span_id = ""
        return True


def install_log_correlation(logger: logging.Logger | None = None) -> TraceCorrelationFilter:
    """Attach a :class:`TraceCorrelationFilter` to ``logger``'s handlers (the root logger by
    default) so records emitted under a span carry ``trace_id`` / ``span_id``. The entrypoint
    calls this once after configuring logging; correlation then applies to any log call made
    while a span is active. Returns the installed filter."""
    target = logger if logger is not None else logging.getLogger()
    filt = TraceCorrelationFilter()
    for handler in target.handlers:
        handler.addFilter(filt)
    return filt

File: go/internal/obs/log.go
package obs

import (
	"context"
	"log/slog"

	"go.opentelemetry.io/otel/trace"
)

// logHandler decorates a slog.Handler so every record emitted while a span is active also
// carries that span's trace_id and span_id. This is the log<->trace correlation seam: it
// lets a backend pivot from a log line to the trace it belongs to (and, on the GCP path,
// Cloud Logging auto-links the two). When no span is active — or tracing is disabled, in
// which case the active span is the framework's no-op span with an invalid context — it
// adds nothing, so logging is unchanged.
type logHandler struct {
	slog.Handler
}

// NewLogHandler wraps base so records gain trace_id/span_id from the active span on the
// record's context. The entrypoint wraps its slog handler with this once; correlation then
// applies to any context-aware log call (slog's *Context methods) made under a span.
func NewLogHandler(base slog.Handler) slog.Handler {
	return logHandler{Handler: base}
}

// Handle adds the active span's ids to the record before delegating. The span comes from
// the record's context, so only context-aware log calls correlate — a deliberate, zero-cost
// design: a log call with no span (or with tracing off) is delegated untouched.
func (h logHandler) Handle(ctx context.Context, rec slog.Record) error {
	if sc := trace.SpanContextFromContext(ctx); sc.IsValid() {
		rec.AddAttrs(
			slog.String("trace_id", sc.TraceID().String()),
			slog.String("span_id", sc.SpanID().String()),
		)
	}
	return h.Handler.Handle(ctx, rec)
}

// WithAttrs and WithGroup preserve the decoration across derived handlers: slog calls these
// to build child handlers (logger.With(...), logger.WithGroup(...)), and without overriding
// them the wrapper would be unwrapped back to the bare base handler, dropping correlation.
func (h logHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
	return logHandler{Handler: h.Handler.WithAttrs(attrs)}
}

func (h logHandler) WithGroup(name string) slog.Handler {
	return logHandler{Handler: h.Handler.WithGroup(name)}
}
